File: today/flatten-binary-tree-to-linked-list.ts
/**
 * https://leetcode.com/problems/flatten-binary-tree-to-linked-list/
 */
export { flatten, TreeNode };

function flatten(root: TreeNode | null): void {
  if (!root) return;

  const nodes: TreeNode[] = [];
  function dfs(node: TreeNode) {
    nodes.push(node);
    if (node.left) dfs(node.left);
    if (node.right) dfs(node.right);
  }

  dfs(root);

  for (let i = 0; i < nodes.length - 1; i++) {
    nodes[i].left = null;
    nodes[i].right = nodes[i + 1];
  }
}

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

File: returned/python/today9/flatten_binary_tree_to_linked_list.py
# https://leetcode.com/problems/flatten-binary-tree-to-linked-list/description/
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        if not root:
            return None

        nodes = []
        def dfs(node: Optional[TreeNode]) -> None:
            nodes.append(node)

            if node.left:
                dfs(node.left)

            if node.right:
                dfs(node.right)
        
        dfs(root)

        for i in range(len(nodes)-1):
            nodes[i].left = None
            nodes[i].right = nodes[i+1]
        
        return nodes[0]